src/app/credExplorer/pagerankTable/aggregation.test.js

```diff
@@ -1,7 +1,422 @@
+        source: nodes.kif,
+        sourceScore: 0.2,
+        connectionScore: 2,
+      },
+      rfk1: {
+        connection: connections.rfk1,
+        source: nodes.kif,
+        sourceScore: 0.2,
+        connectionScore: 1,
+      },
+    };
+    const scoredConnectionsArray = [
+      scoredConnections.loop5,
+      scoredConnections.rfz4,
+      scoredConnections.rfk1,
+      scoredConnections.rbk2,
+      scoredConnections.kfr3,
+    ];
+    return {
+      nodes,
+      edges,
+      nodeTypes,
+      nodeTypesArray,
+      edgeTypes,
+      edgeTypesArray,
+      adjacencies,
+      connections,
+      scoredConnections,
+      scoredConnectionsArray,
+    };
+  }
+  describe("aggregateByNodeType", () => {
+    it("puts every connection in an aggregation", () => {
+      const {nodeTypesArray, scoredConnectionsArray} = example();
+      const aggregations = aggregateByNodeType(
+        scoredConnectionsArray,
+        nodeTypesArray
+      );
+      let connectionsFound = 0;
+      for (const aggregation of aggregations) {
+        connectionsFound += aggregation.connections.length;
+      }
+      expect(connectionsFound).toEqual(scoredConnectionsArray.length);
+    });
+    it("groups connections by node type", () => {
+      const {nodeTypesArray, scoredConnectionsArray} = example();
+      const aggregations = aggregateByNodeType(
+        scoredConnectionsArray,
+        nodeTypesArray
+      );
+      for (const aggregation of aggregations) {
+        for (const connection of aggregation.connections) {
+          expect(
+            NodeAddress.hasPrefix(
+              connection.source,
+              aggregation.nodeType.prefix
+            )
+          ).toBe(true);
+        }
+      }
+    });
+    it("when multiple node types match, it uses the most specific one", () => {
+      const {nodeTypes, scoredConnections} = example();
+      const aggregations = aggregateByNodeType(
+        [scoredConnections.rfz4],
+        [nodeTypes.empty, nodeTypes.zap]
+      );
+      expect(aggregations).toHaveLength(1);
+      // It used the zap nodeType rather than the empty nodeType b/c it was more specific
+      expect(aggregations[0].nodeType).toEqual(nodeTypes.zap);
+    });
+    it("summarizes the group size appropriately", () => {
+      const {nodeTypesArray, scoredConnectionsArray} = example();
+      const aggregations = aggregateByNodeType(
+        scoredConnectionsArray,
+        nodeTypesArray
+      );
+      for (const aggregation of aggregations) {
+        expect(aggregation.summary.size).toEqual(
+          aggregation.connections.length
+        );
+      }
+    });
+    it("summarizes the group score appropriately", () => {
+      const {nodeTypesArray, scoredConnectionsArray} = example();
+      const aggregations = aggregateByNodeType(
+        scoredConnectionsArray,
+        nodeTypesArray
+      );
+      for (const aggregation of aggregations) {
+        const totalScore = aggregation.connections.reduce(
```

Comment:
Nit incoming, but one that I care about somewhat. You’ve a couple times
now written code of the form

```js
        const totalScore = aggregation.connections.reduce(
          (val, connection) => val + connection.connectionScore,
          0
        );
```

, which can be rewritten as

```js
        const scores = aggregation.connections.map((x) => x.connectionScore);
        const totalScore = scores.reduce((a, b) => a + b);
```

. I posit that the latter is better, both for “ivory-tower” “purity”
reasons\* and for practical reasons. The practical reason is simply the
following: while it is not _difficult_ to rederive at each occasion
whether the accumulator or the next item comes as the first argument to
the callback, it requires active thought each time that you read the
code (for me, at least, and I’m not alone). When your accumulator type
is different from your element type, the order matters. When you map
first so that everything is homogeneous, it doesn’t.\*\*

I think that this is consonant with the principle that code is read much
more often than it’s written; the obviously correct version is better
than the slightly-nonobviously correct version.

Another practical reason is that it’s impossible to get wrong if you map
first and then reduce. If you got the arguments wrong here, you’d get
`0` for an empty array or `"[object Object]undefined"` for a nonempty
array. Sure, tests will catch it, but why bother tracking down that bug?

\* The ivory-tower reason is roughly that when you perform a true
_reduction_, as opposed to a fold, you have much more freedom in terms
of execution strategy. For instance, it is trivial to parallelize a
reduction, but it is impossible to parallelize a fold. This really does
matter in principle (you sacrifice O(log *n*) for O(*n*)), but in
practice this code is not parallel, and if we want to make it parallel
then we can trivially make the transformation.

\*\* This is true even if the operator is non-commutative—ask me for
details if you care.

So—I care about this enough to write my own code in my preferred style,
and I want to let you know where I’m coming from. But if you have any
remotely good reasons for favoring the former style in any particular
case then I’m happy to let it slide.

\</nit> :-)

Reply:
Yep, this is a great suggestion. I got confused about the reduce argument order several times while writing the code. I'm surprised that reduces can be parallelized when non-commutative, if either of us remember tomorrow I want to hear why.